File: Backend-Folder/backend/app/schemas/audit.py
from typing import Optional, List
from pydantic import BaseModel

class AuditListResponse(BaseModel):
    success: bool = True
    count: int
    data: List[dict]

class AuditVerifyResponse(BaseModel):
    success: bool = True
    verified: bool = True
    totalBlocks: int
    ledgerState: str = "IMMUTABLE_SYNCED"
    latestBlockHash: Optional[str] = None

File: Backend-Folder/server/routes/audit.ts
import { Router } from 'express';
import { db } from '../db';

const router = Router();

// GET /api/audit-trail - Cryptographic ledger
router.get('/', (req, res) => {
  try {
    const logs = db.getAuditLogs();
    res.json({ success: true, count: logs.length, data: logs });
  } catch (err: any) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// POST /api/audit-trail/verify - Verify signature chain integrity
router.post('/verify', (req, res) => {
  try {
    const logs = db.getAuditLogs();
    // Validate that hashes are non-empty and formatted
    const isValid = logs.every((l) => l.hash.startsWith('sha256:'));
    res.json({
      success: true,
      verified: isValid,
      totalBlocks: logs.length,
      ledgerState: 'IMMUTABLE_SYNCED',
      latestBlockHash: logs[0]?.hash,
    });
  } catch (err: any) {
    res.status(500).json({ success: false, error: err.message });
  }
});

export default router;
